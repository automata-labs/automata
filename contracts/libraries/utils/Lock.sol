// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import "hardhat/console.sol";

abstract contract Lock {
    bool public unlocked = true; // unlocked at creation

    modifier lock() {
        require(unlocked, "LOK");
        unlocked = false;
        _;
        unlocked = true;
    }
}