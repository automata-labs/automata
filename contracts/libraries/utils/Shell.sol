// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

import "../data/Slot.sol";
import "../../interfaces/IKernel.sol";

library Shell {
    function fetch(IKernel kernel, address underlying, address owner) internal view returns (Slot.Data memory) {
        return kernel.get(keccak256(abi.encode(underlying, owner)));
    }

    function modify(IKernel kernel, address underlying, address owner, int128 deltaX, int128 deltaY) internal {
        kernel.update(keccak256(abi.encode(underlying, owner)), deltaX, deltaY);
    }

    function move(IKernel kernel, address underlying, address from, address to, uint128 x, uint128 y) internal {
        kernel.transfer(keccak256(abi.encode(underlying, from)), keccak256(abi.encode(underlying, to)), x, y);
    }
}
