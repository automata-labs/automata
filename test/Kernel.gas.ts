import { use } from 'chai';
import { ethers, waffle } from 'hardhat';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';

import { deploy, snapshotGasCost } from './shared/utils';
import { Kernel } from '../typechain';

use(jestSnapshotPlugin());

const { loadFixture } = waffle;

describe('Kernel', async () => {
  let abi = new ethers.utils.AbiCoder();
  let wallet;
  let kernel: Kernel;

  const fixture = async () => {
    ;([wallet] = await ethers.getSigners());
    kernel = (await deploy('Kernel')) as Kernel;
  };

  beforeEach(async () => {
    await loadFixture(fixture);
  });

  describe('#write', async () => {
    it('gas first write', async () => {
      const key = ethers.utils.keccak256(abi.encode(['address'], [wallet.address]));
      await snapshotGasCost(kernel.write(key, 1, 2));
    });
    it('gas write', async () => {
      const key = ethers.utils.keccak256(abi.encode(['address'], [wallet.address]));
      await kernel.write(key, 1, 2);
      await snapshotGasCost(kernel.write(key, 2, 3));
    });
  });

  describe('#update', async () => {
    it('gas update', async () => {
      const key = ethers.utils.keccak256(abi.encode(['address'], [wallet.address]));
      await kernel.write(key, 1, 1);
      await snapshotGasCost(kernel.update(key, 1, 1));
    });
  });
});
