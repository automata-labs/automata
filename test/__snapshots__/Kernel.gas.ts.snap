// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kernel #update gas update 1`] = `37373`;

exports[`Kernel #write gas first write 1`] = `52984`;

exports[`Kernel #write gas write 1`] = `35884`;
